### ISSUE TITLE
fix: Patching lazily-linked natives is broken in mixed mode

### ISSUE THREAD

**Issue by kmillikin** [shorebirdtech/shorebird#1448]
Native calls have (1) "what" native function to call and (2) "how" to call it which is one of bootstrap native, no scope native, or auto scope native.

Lazily linked native calls have LinkNativeCall as "what" to call and bootstrap native as "how" to call.  Then they call LinkNativeCall that replaces the call site with a different what and how.

In mixed mode, we see a call in the kernel-service snapshot that has been set to call LinkNativeCall as an auto scope native.  This won't work (LinkNativeCall only works when called as a bootstrap native).

It's not clear how this is happening.  (1) we might be compiling it incorrectly in the first place as a non-lazily linked native.  (2) we might be patching it incorrectly before writing out the snaphot.

It appears in the kernel-service snapshot.

**Comment by kmillikin** [comment]
The command to compile the kernel service snapshot is (run from the build directory, e.g. `xcodebuild/DebugARM64`):

```
./dart --deterministic --packages=../../.dart_tool/package_config.json --snapshot=gen/kernel-service.dart.snapshot \
    --snapshot-depfile=gen/kernel-service.dart.snapshot.d --snapshot-kind=app-jit \
    --dfe=../../utils/kernel-service/NEVER_LOADED gen/utils/kernel-service/kernel-service_snapshot.dart.dill \
    --train file:///Users/kmillikin/dart-sdk/sdk/pkg/vm/bin/kernel_service.dart
```

Make sure to use the compiled `dart` binary from that directory and not one from your PATH, because otherwise the Kernel version might not match.

**Comment by kmillikin** [comment]
Running a simple Dart program that's just:

```dart
main() {
  print("test");
}
```

will trigger the failure in trying to patch a native call in the function `dart:io__NamespaceImpl@14069316__getDefault@14069316` from the kernel service isolate.

**Comment by kmillikin** [comment]
So I figured out what's happening here:

The serializer replaces the native entry to call with `LinkNativeCall`, because it doesn't know what the addresses of any of the native entries will be when it deserializes the snapshot.  This turns the call site into a lazily-linked one.  Therefore, the serializer also replaces the stub that wraps the native with the `CallBootstrapNative` stub (the one that must be used for `LinkNativeCall`).

I have modified the VM so the direct call to a specific stub is instead a call to a stub found at a specific thread-relative offset.  So therefore the serializer needs to replace whatever offset it finds with the offset of the `CallBootstrapNative` stub when it replaces the native entry with `LinkNativeCall`.
